parsers/ELEXON.py

```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python3
+# coding=utf-8
+
+"""
+Parser that uses the ELEXON API to return the following data types.
+
+Production
+Exchanges
+
+Documentation:
+https://www.elexon.co.uk/wp-content/uploads/2017/06/
+bmrs_api_data_push_user_guide_v1.1.pdf
+"""
+
+import os
+import arrow
+import logging
+import requests
+import datetime as dt
+
+ELEXON_ENDPOINT = 'https://api.bmreports.com/BMRS/{}/v1'
+
+FUELS = {
+    'production': {
+        'gas': [4, 11],
+        'oil': 5,
+        'coal': 6,
+        'nuclear': 7,
+        'wind': 8,
+        'hydro': 10,
+        'unknown': 12,
+        'biomass': 17
+    },
+    'storage': {
+        'hydro': 9
+    }
+}
+
+EXCHANGES = {
+    'FR->GB': 3,
+    'GB-NIR->IE': 4,
+    'GB->NL': 5,
+    'GB->IE': 6
+}
+
+EXPECTED_FIELDS = {
+    'FUELINST': 18,
+    'INTERFUELHH': 7
+}
+
+
+def query_ELEXON(report, session, params, target_datetime=None,
+                 span=(-48, 24)):
+    if target_datetime is None:
+        target_datetime = arrow.utcnow().to('Europe/London')
+    else:
+        # make sure we have an arrow object
+        target_datetime = arrow.get(target_datetime).to('Europe/London')
+
+    params['FromDate'] = target_datetime.replace(
+        hours=span[0]).format('YYYY-MM-DD')
+    params['ToDate'] = target_datetime.replace(
+        hours=span[1]).format('YYYY-MM-DD')
+
+    if 'ELEXON_TOKEN' not in os.environ:
+        raise Exception('No ELEXON_TOKEN found! Please add it to secrets.env!')
+    params['APIKey'] = os.environ['ELEXON_TOKEN']
+    return session.get(ELEXON_ENDPOINT.format(report), params=params)
+
+
+def query_exchange(session, target_datetime=None):
+    params = {
+        'ServiceType': 'csv'
+    }
+    response = query_ELEXON('INTERFUELHH', session, params, target_datetime)
+    response = response.content.decode()
+    return response
+
+
+def query_production(session, target_datetime=None):
+    params = {
+        'ServiceType': 'csv'
+    }
+    response = query_ELEXON('FUELINST', session, params, target_datetime)
+    response = response.content.decode()
+    return response
+
+
+def parse_exchange(zone_key1, zone_key2, csv_text, target_datetime=None,
+                   logger=logging.getLogger(__name__)):
+    if not csv_text:
+        return None
+
+    sorted_zone_keys = sorted([zone_key1, zone_key2])
+
+    exchange = '->'.join(sorted_zone_keys)
+    data_points = list()
+    lines = csv_text.split('\n')
+
+    # check field count in report is as expected
+    field_count = len(lines[1].split(','))
+    if field_count != EXPECTED_FIELDS['INTERFUELHH']:
+        logger.warning(
+            'Expected {} fields in INTERFUELHH report, got {}'.format(
+                EXPECTED_FIELDS['INTERFUELHH'], field_count
+            )
+        )
+
+    for line in lines[1:-1]:
+        fields = line.split(',')
+
+        # create the datetime
+        datetime = datetime_from_date_sp(fields[1], fields[2])
```

Comment:
What does fields[2] represent? Is that an offset from fields[1] in minutes?

Reply:
fields[2] is the settlement period. So there are 48 settlement periods each day, each is 30 minutes long. The `datetime_from_date_sp` function contains the logic to convert a settlement date and settlement period to a local datetime.

Reply:
(this conversion does not yet make considerations for the time change when one sees settlement periods 49 and 50)